File: test/Zametek.Maths.Graphs.Compilers.Tests/Engines/DependencyResolutionTests.cs
using Shouldly;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace Zametek.Maths.Graphs.Tests
{
    // Correctness tests for the transitive reducer's shared-ancestor handling and the
    // strong-dependency walk's dummy-chain resolution - the paths whose recursion was
    // replaced with iterative, visited-set traversals.
    public class DependencyResolutionTests
    {
        [Fact]
        public void VertexGraphBuilder_GivenSharedAncestorAcrossTwoSinks_ThenTransitiveReductionRemovesRedundantEdgesFromBoth()
        {
            var builder = new VertexGraphBuilder<int, int, int, IActivity<int, int, int>>(new NextIdGenerator<int>(0));

            // A -> B, and two sinks (D1, D2) each depending on BOTH A (directly, redundantly)
            // and B. A is a shared ancestor of both sinks; the direct A->D1 and A->D2 edges are
            // redundant and must both be removed (guards the reducer's single shared visited set).
            builder.AddActivity(new Activity<int, int, int>(1, 1)).ShouldBeTrue();        // A
            builder.AddActivity(new Activity<int, int, int>(2, 1), [1]).ShouldBeTrue();    // B  dep A
            builder.AddActivity(new Activity<int, int, int>(3, 1), [1, 2]).ShouldBeTrue(); // D1 dep A, B
            builder.AddActivity(new Activity<int, int, int>(4, 1), [1, 2]).ShouldBeTrue(); // D2 dep A, B

            builder.TransitiveReduction().ShouldBeTrue();

            builder.ActivityDependencyIds(3).ShouldBe([2]);
            builder.ActivityDependencyIds(4).ShouldBe([2]);
            builder.ActivityDependencyIds(2).ShouldBe([1]);
        }

        [Fact]
        public void VertexGraphBuilder_GivenRealRootBehindDummyDiamond_ThenStrongActivityDependencyIdsResolvesRealRoot()
        {
            var builder = new VertexGraphBuilder<int, int, int, IActivity<int, int, int>>(new NextIdGenerator<int>(0));

            // A (real) sits behind two zero-duration (dummy) links B and C, which D depends on.
            builder.AddActivity(new Activity<int, int, int>(1, 1)).ShouldBeTrue();        // A real
            builder.AddActivity(new Activity<int, int, int>(2, 0), [1]).ShouldBeTrue();    // B dummy dep A
            builder.AddActivity(new Activity<int, int, int>(3, 0), [1]).ShouldBeTrue();    // C dummy dep A
            builder.AddActivity(new Activity<int, int, int>(4, 1), [2, 3]).ShouldBeTrue(); // D real dep B, C

            List<int> strong = builder.StrongActivityDependencyIds(4);

            // A is reachable via two dummy paths; it must be found (once, as a set) and the
            // dummies themselves excluded.
            strong.Distinct().ShouldBe([1]);
            strong.ShouldNotContain(2);
            strong.ShouldNotContain(3);
        }

        [Fact]
        public void ArrowGraphBuilder_GivenRealRootBehindDummyDiamond_ThenStrongActivityDependencyIdsResolvesRealRoot()
        {
            var builder = new ArrowGraphBuilder<int, int, int, IActivity<int, int, int>>(
                new NextIdGenerator<int>(100),
                new NextIdGenerator<int>(0));

            builder.AddActivity(new Activity<int, int, int>(1, 1)).ShouldBeTrue();        // A real
            builder.AddActivity(new Activity<int, int, int>(2, 0), [1]).ShouldBeTrue();    // B dummy dep A
            builder.AddActivity(new Activity<int, int, int>(3, 0), [1]).ShouldBeTrue();    // C dummy dep A
            builder.AddActivity(new Activity<int, int, int>(4, 1), [2, 3]).ShouldBeTrue(); // D real dep B, C

            List<int> strong = builder.StrongActivityDependencyIds(4);

            strong.Distinct().ShouldBe([1]);
            strong.ShouldNotContain(2);
            strong.ShouldNotContain(3);
        }
    }
}

File: rust/compilers/tests/dependency_resolution_tests.rs
//! Ports of `DependencyResolutionTests.cs`: the transitive reducer's shared-
//! ancestor handling and the strong-dependency walk's dummy-chain resolution,
//! exercised through the public builder API.

use indexmap::IndexSet;
use zametek_maths_graphs_compilers::{ArrowGraphBuilder, NextIdGenerator, VertexGraphBuilder};
use zametek_maths_graphs_primitives::DependentActivity;

type Act = DependentActivity<i32, i32, i32>;

#[test]
fn vertex_shared_ancestor_across_two_sinks_then_reduction_removes_redundant_edges_from_both() {
    // A -> B, and two sinks each depending on BOTH A (redundantly) and B.
    let mut builder = VertexGraphBuilder::<i32, i32, i32>::new(NextIdGenerator::new(0));
    assert!(builder.add_activity(Act::new(1, 1))); // A
    builder.add_activity_with_dependencies(Act::new(2, 1), IndexSet::from([1])); // B dep A
    builder.add_activity_with_dependencies(Act::new(3, 1), IndexSet::from([1, 2])); // D1 dep A, B
    builder.add_activity_with_dependencies(Act::new(4, 1), IndexSet::from([1, 2])); // D2 dep A, B

    assert!(builder.transitive_reduction());

    assert_eq!(builder.activity_dependency_ids(3), vec![2]);
    assert_eq!(builder.activity_dependency_ids(4), vec![2]);
    assert_eq!(builder.activity_dependency_ids(2), vec![1]);
}

#[test]
fn vertex_real_root_behind_dummy_diamond_then_strong_dependency_resolves_real_root() {
    // A (real) behind two zero-duration links B and C, which D depends on.
    let mut builder = VertexGraphBuilder::<i32, i32, i32>::new(NextIdGenerator::new(0));
    assert!(builder.add_activity(Act::new(1, 1)));
    builder.add_activity_with_dependencies(Act::new(2, 0), IndexSet::from([1]));
    builder.add_activity_with_dependencies(Act::new(3, 0), IndexSet::from([1]));
    builder.add_activity_with_dependencies(Act::new(4, 1), IndexSet::from([2, 3]));

    let mut strong = builder.strong_activity_dependency_ids(4);
    strong.sort();
    strong.dedup();
    assert_eq!(strong, vec![1]);
}

#[test]
fn arrow_real_root_behind_dummy_diamond_then_strong_dependency_resolves_real_root() {
    let mut builder =
        ArrowGraphBuilder::<i32, i32, i32>::new(NextIdGenerator::new(100), NextIdGenerator::new(0));
    assert!(builder.add_activity(Act::new(1, 1)));
    builder.add_activity_with_dependencies(Act::new(2, 0), IndexSet::from([1]));
    builder.add_activity_with_dependencies(Act::new(3, 0), IndexSet::from([1]));
    builder.add_activity_with_dependencies(Act::new(4, 1), IndexSet::from([2, 3]));

    let mut strong = builder.strong_activity_dependency_ids(4);
    strong.sort();
    strong.dedup();
    assert_eq!(strong, vec![1]);
}
